### PR TITLE
Feature/tao 3140 expose category mapping

### PR DESCRIPTION
**Pull request by krampstudio**
I've moved everything to taoItems (we use categories in the QTI Test, but the main purpose to _categorize_ resources isn't directly linked to QTI).
There's a new property to capture the exposition state of a property.

In the class properties form, the "tag" icon let's you switch the exposition state. Once a category is exposed, we don't create a mapping but expose the properties values as categories (after a sanitization step). So only some properties are eligible to categorization. 

@jsconan could you please check the frontend side.

### CODE REVIEW

File: models/classes/CategoryService.php [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    const ITEM_CLASS_URI  = 'http://www.tao.lu/Ontologies/TAOItem.rdf#Item';
+    const EXPOSE_PROP_URI = 'http://www.tao.lu/Ontologies/TAOItem.rdf#ExposeCategory';
+    const SUPPORTED_WIDGET_URIS = [
+        'http://www.tao.lu/datatypes/WidgetDefinitions.rdf#TextBox',
+        'http://www.tao.lu/datatypes/WidgetDefinitions.rdf#CheckBox',
+        'http://www.tao.lu/datatypes/WidgetDefinitions.rdf#RadioBox',
+        'http://www.tao.lu/datatypes/WidgetDefinitions.rdf#ComboBox',
+        'http://www.tao.lu/datatypes/WidgetDefinitions.rdf#TreeBox'
+    ];
+
```
Comment (antoinerobin) [inline models/classes/CategoryService.php]: Arrays are not allowed in class constants

File: views/js/component/category/switch.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    'tpl!taoItems/component/category/switch'
+], function($, _, __, component, switchTpl){
+    'use strict';
+
+    /**
+     * The titles based on current exposition state
+     */
+    var titles = {
+        expose : __('Expose the property value to create automated categories.'),
+        unexpose : __('The property is exposed to category addtion.&#13Remove exposition ?')
```
Comment (jsconan) [inline views/js/component/category/switch.js]: There is a typo here: `addtion`=> `addition` ?

File: models/classes/CategoryService.php (hunk 1 above)
Comment (krampstudio) [inline models/classes/CategoryService.php]: It's introduced in PHP5.6. What version do we support ?

Comment (antoinerobin) [inline models/classes/CategoryService.php]: Unfortunately php5.5

Comment (krampstudio) [inline models/classes/CategoryService.php]: Fixed

File: views/js/component/category/switch.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+ * Copyright (c) 2016 (original work) Open Assessment Technologies SA;
+ */
+
+/**
+ * A component to switch the exposition state for categories of a property
+ *
+ * @author Bertrand Chevrier <bertrand@taotesting.com>
+ */
+define([
+    'jquery',
```
Comment (jsconan) [inline views/js/component/category/switch.js]: Not an issue, juste a though. jQuery does not seem to be directly used by this module. Is it still needed to require it ?

File: views/js/controller/items/editItemClass.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    'jquery',
+    'lodash',
+    'uri',
+    'taoItems/component/category/switch',
+    'taoItems/provider/category',
+    'ui/feedback'
+], function($, _, uriUtil, categorySwitch, categoryProvider, feedback) {
+    'use strict';
+
+    var provider  = categoryProvider();
```
Comment (jsconan) [inline views/js/controller/items/editItemClass.js]: The is no JSDoc in this module. Is it ok ?

**Comment by antoinerobin** [comment]
ItemCategory service won't be set at install time. Updater correctly register it to the serviceManager but nothing is done when we install the extension

File: views/js/controller/items/editItemClass.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                        if(success){
+                            self.setExposed(value);
+                        }
+                    })
+                    .catch(handleError);
+            });
+    };
+
+
+    var indexCategoryController = {
```
Comment (jsconan) [inline views/js/controller/items/editItemClass.js]: Not an issue, just a thought. Is it needed to create a variable for this object ? It is not referenced, except for the return statement.

File: views/js/controller/items/index.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014 (original work) Open Assessment Technologies SA;
+ */
+define([
+],
+function(){
```
Comment (jsconan) [inline views/js/controller/items/index.js]: The require list is empty, maybe you could just define the function, isn't it ?

File: views/js/test/provider/category/test.js [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+    QUnit.asyncTest('success', function(assert) {
+        var p;
+        var provider = categoryProvider( testConfig );
+
+        QUnit.expect(4);
+
+        p = provider.getExposedsByClass('classUri');
+
+        assert.ok(p instanceof Promise, 'The method returns a Promise');
```
Comment (jsconan) [inline views/js/test/provider/category/test.js]: The Promise polyfill has not been required

**Comment by krampstudio** [comment]
Added the installer

File: scripts/install/RegisterCategoryService.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2016 (original work) Open Assessment Technologies SA;
+ *
+ *
+ */
+
+namespace oat\taoItems\scripts\install;
+
+use common_ext_action_InstallAction as InstallAction;
```
Comment (antoinerobin) [inline scripts/install/RegisterCategoryService.php]: you can use the new oat\oatbox\extension\InstallAction class

**Comment by krampstudio** [comment]
everything fixed

**Comment by jsconan** [comment]
Front end side is Ok

File: models/classes/CategoryService.php [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    /**
+     * Expose or not a property
+     *
+     * @param RdfProperty $property the property to check
+     * @param bool        $value    true if exposed
+     *
+     * @return void
+     */
+    public function exposeCategory(RdfProperty $property, bool $value)
```
Comment (antoinerobin) [inline models/classes/CategoryService.php]: you can't use type hitting with primitive type as boolean

**Comment by krampstudio** [comment]
I'll replace them by static fields

Bertrand Chevrier
Software craftsman | OAT S.A.
Be free, be creative | http://taotesting.com

**Pull request merged**